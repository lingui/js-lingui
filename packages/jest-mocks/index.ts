export function getConsoleMockCalls({ mock }: jest.MockInstance<any, any>) {
  if (!mock.calls.length) return
  return mock.calls.map((call) => call[0]).join("\n")
}

export function mockConsole(
  testCase: (console: jest.Mocked<Console>) => any,
  mock = {}
) {
  function restoreConsole() {
    global.console = originalConsole
  }

  const originalConsole = global.console

  const defaults = {
    log: jest.fn(),
    warn: jest.fn(),
    error: jest.fn(),
  }

  global.console = {
    ...defaults,
    ...mock,
  } as any

  let result
  try {
    result = testCase(global.console as jest.Mocked<Console>)
  } catch (e) {
    restoreConsole()
    throw e
  }

  if (result && typeof result.then === "function") {
    return result.then(restoreConsole).catch((e) => {
      restoreConsole()
      throw e
    })
  } else {
    restoreConsole()
    return result
  }
}

export function mockEnv(env, testCase) {
  const oldEnv = process.env.NODE_ENV
  process.env.NODE_ENV = env

  try {
    testCase()
  } catch (e) {
    process.env.NODE_ENV = oldEnv
    throw e
  }

  process.env.NODE_ENV = oldEnv
}
