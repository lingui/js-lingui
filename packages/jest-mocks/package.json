{
  "private": true,
  "name": "@lingui/jest-mocks",
  "version": "0.0.0-managed-by-release-script",
  "description": "Mocks for Jest",
  "main": "index.js",
  "author": {
    "name": "Tomáš Ehrlich",
    "email": "tomas.ehrlich@gmail.com"
  },
  "license": "MIT",
  "keywords": [
    "jest",
    "testing",
    "mock"
  ],
  "repository": {
    "type": "git",
    "url": "https://github.com/lingui/js-lingui.git"
  },
  "bugs": {
    "url": "https://github.com/lingui/js-lingui/issues"
  },
  "engines": {
    "node": ">=6.0"
  },
  "files": [
    "LICENSE",
    "README.md",
    "index.js",
    "*.js.flow"
  ]
}
