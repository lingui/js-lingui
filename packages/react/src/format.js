// @flow
import { cloneElement } from "react"

// match <0>paired</0> and <1/> unpaired tags
const tagRe = /<(\d+)>(.*)<\/\1>|<(\d+)\/>/
const nlRe = /(?:\r\n|\r|\n)/g

/**
 * `formatElements` - parse string and return tree of react elements
 *
 * `value` is string to be formatted with <0>Paired<0/> or <0/> (unpaired)
 * placeholders. `elements` is a array of react elements which indexes
 * correspond to element indexes in formatted string
 */
function formatElements(
  value: string,
  elements: Array<React$Element<any>> = []
): string | Array<any> {
  // TODO: warn if there're any unprocessed elements
  // TODO: warn if element at `index` doesn't exist

  const parts = value.replace(nlRe, "").split(tagRe)

  // no inline elements, return
  if (parts.length === 1) return value

  const tree = []

  const before = parts.shift()
  if (before) tree.push(before)

  for (const [index, children, after] of getElements(parts)) {
    const element = elements[index]
    tree.push(
      cloneElement(
        element,
        { key: index },

        // format children for pair tags
        // unpaired tags might have children if it's a component passed as a variable
        children ? formatElements(children, elements) : element.props.children
      )
    )

    if (after) tree.push(after)
  }

  return tree
}

/*
 * `getElements` - return array of element indexes and element childrens
 *
 * `parts` is array of [pairedIndex, children, unpairedIndex, textAfter, ...]
 * where:
 * - `pairedIndex` is index of paired element (undef for unpaired)
 * - `children` are children of paired element (undef for unpaired)
 * - `unpairedIndex` is index of unpaired element (undef for paired)
 * - `textAfter` is string after all elements (empty string, if there's nothing)
 *
 * `parts` length is always multiply of 4
 *
 * Returns: Array<[elementIndex, children, after]>
 */
function getElements(parts) {
  if (!parts.length) return []

  const [paired, children, unpaired, after] = parts.slice(0, 4)

  return [[parseInt(paired || unpaired), children || "", after]].concat(
    getElements(parts.slice(4, parts.length))
  )
}

export { formatElements }
