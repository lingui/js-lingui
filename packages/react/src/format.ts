import React from "react"

// match <tag>paired</tag> and <tag/> unpaired tags
const tagRe = /<([a-zA-Z0-9]+)>(.*?)<\/\1>|<([a-zA-Z0-9]+)\/>/
const nlRe = /(?:\r\n|\r|\n)/g

// For HTML, certain tags should omit their close tag. We keep a whitelist for
// those special-case tags.
const voidElementTags = {
  area: true,
  base: true,
  br: true,
  col: true,
  embed: true,
  hr: true,
  img: true,
  input: true,
  keygen: true,
  link: true,
  meta: true,
  param: true,
  source: true,
  track: true,
  wbr: true,
  menuitem: true,
}

/**
 * `formatElements` - parse string and return tree of react elements
 *
 * `value` is string to be formatted with <tag>Paired<tag/> or <tag/> (unpaired)
 * placeholders. `elements` is a array of react elements which indexes
 * correspond to element indexes in formatted string
 */
function formatElements(
  value: string,
  elements: { [key: string]: React.ReactElement } = {}
): string | Array<React.ReactElement | string> {
  const uniqueId = makeCounter(0, "$lingui$")
  const parts = value.replace(nlRe, "").split(tagRe)

  // no inline elements, return
  if (parts.length === 1) return value

  const tree: Array<React.ReactElement | string> = []

  const before = parts.shift()
  if (before) tree.push(before)

  for (const [index, children, after] of getElements(parts)) {
    let element = typeof index !== "undefined" ? elements[index] : undefined

    if (
      !element ||
      (voidElementTags[element.type as keyof typeof voidElementTags] &&
        children)
    ) {
      if (!element) {
        console.error(
          `Can't use element at index '${index}' as it is not declared in the original translation`
        )
      } else {
        console.error(
          `${element.type} is a void element tag therefore it must have no children`
        )
      }

      // ignore problematic element but push its children and elements after it
      element = React.createElement(React.Fragment)
    }

    if (Array.isArray(element)) {
      element = React.createElement(React.Fragment, {}, element)
    }

    tree.push(
      React.cloneElement(
        element,
        { key: uniqueId() },

        // format children for pair tags
        // unpaired tags might have children if it's a component passed as a variable
        children ? formatElements(children, elements) : element.props.children
      )
    )

    if (after) tree.push(after)
  }

  return tree
}

/*
 * `getElements` - return array of element indices and element children
 *
 * `parts` is array of [pairedIndex, children, unpairedIndex, textAfter, ...]
 * where:
 * - `pairedIndex` is index of paired element (undef for unpaired)
 * - `children` are children of paired element (undef for unpaired)
 * - `unpairedIndex` is index of unpaired element (undef for paired)
 * - `textAfter` is string after all elements (empty string, if there's nothing)
 *
 * `parts` length is always a multiple of 4
 *
 * Returns: Array<[elementIndex, children, after]>
 */
function getElements(
  parts: string[]
): Array<readonly [string | undefined, string, string | undefined]> {
  if (!parts.length) return []

  const [paired, children, unpaired, after] = parts.slice(0, 4)

  const triple = [paired || unpaired, children || "", after] as const
  return [triple].concat(getElements(parts.slice(4, parts.length)))
}

const makeCounter =
  (count = 0, prefix = "") =>
  () =>
    `${prefix}_${count++}`

export { formatElements }
