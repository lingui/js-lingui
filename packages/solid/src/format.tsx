import { JSX } from "solid-js"

// match <tag>paired</tag> and <tag/> unpaired tags
const tagRe = /<([a-zA-Z0-9]+)>(.*?)<\/\1>|<([a-zA-Z0-9]+)\/>/
const nlRe = /(?:\r\n|\r|\n)/g

// For HTML, certain tags should omit their close tag. We keep a whitelist for
// those special-case tags.
const voidElementTags = {
  area: true,
  base: true,
  br: true,
  col: true,
  embed: true,
  hr: true,
  img: true,
  input: true,
  keygen: true,
  link: true,
  meta: true,
  param: true,
  source: true,
  track: true,
  wbr: true,
  menuitem: true,
}

function appendChildren(parent: HTMLElement, children: JSX.Element) {
  if (children == null)
    return
  if (Array.isArray(children)) {
    for (const node of (children as JSX.ArrayElement)) {
      appendChildren(parent, node)
    }
  } else if (children instanceof Node) {
    parent.appendChild(children as Node)
  } else if (typeof children === "string") {
    parent.appendChild(document.createTextNode(children))
  } else {
    console.error(`Invalid children type: ${typeof children}`, children)
  }
}

/**
 * `formatElements` - parse string and return tree of html elements
 *
 * `value` is string to be formatted with <tag>Paired<tag/> or <tag/> (unpaired)
 * placeholders. `elements` is a array of  html elements which indexes
 * correspond to element indexes in formatted string
 */
function formatElements(
  value: string,
  elements: { [key: string]: JSX.Element } = {}
): JSX.Element {
  const parts = value.replace(nlRe, "").split(tagRe)

  // no inline elements, return
  if (parts.length === 1) return value

  const tree: Array<JSX.Element | string> = []

  const before = parts.shift()
  if (before) tree.push(before)

  for (const [index, children, after] of getElements(parts)) {
    let element = typeof index !== "undefined" ? elements[index] : undefined

    if (!element) {
      console.error(`Can't use element at index '${index}' as it is not declared in the original translation`)
      // ignore problematic element but push its children and elements after it
      element = <></>
    } else if (children && (element instanceof Node) && (element as Node).nodeName.toLowerCase() in voidElementTags) {
      console.error(`${(element as Node).nodeName} is a void element tag therefore it must have no children`)
      // ignore problematic element but push its children and elements after it
      element = <></>
    }

    if (children) {
      if (element instanceof HTMLElement) {
        appendChildren(element as HTMLElement, formatElements(children, elements))
      } else {
        if (!Array.isArray(element))
          console.error(`Element is not HTMLElement, but have children:`, element)
        element = <>{element}{formatElements(children, elements)}</>
      }
    }

    tree.push(element)

    if (after) tree.push(after)
  }

  return tree.length === 1 ? tree[0]! : tree
}

/*
 * `getElements` - return array of element indices and element children
 *
 * `parts` is array of [pairedIndex, children, unpairedIndex, textAfter, ...]
 * where:
 * - `pairedIndex` is index of paired element (undef for unpaired)
 * - `children` are children of paired element (undef for unpaired)
 * - `unpairedIndex` is index of unpaired element (undef for paired)
 * - `textAfter` is string after all elements (empty string, if there's nothing)
 *
 * `parts` length is always a multiple of 4
 *
 * Returns: Array<[elementIndex, children, after]>
 */
function getElements(
  parts: string[]
): Array<readonly [string | undefined, string, string | undefined]> {
  if (!parts.length) return []

  const [paired, children, unpaired, after] = parts.slice(0, 4)

  const triple = [paired || unpaired, children || "", after] as const
  return [triple].concat(getElements(parts.slice(4, parts.length)))
}

export { formatElements }
