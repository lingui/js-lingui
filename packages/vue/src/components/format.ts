import { h, type Slot } from "vue"

//

// match <tag>paired</tag> and <tag/> unpaired tags
const tagRe = /<([a-zA-Z0-9]+)>(.*?)<\/\1>|<([a-zA-Z0-9]+)\/>/u
const nlRe = /(?:\r\n|\r|\n)/gu

type Element = ReturnType<typeof h> | ReturnType<Slot>
type Node = Element | string | undefined

/*
 * `getElements` - return array of element indices and element children
 *
 * `parts` is array of [pairedIndex, children, unpairedIndex, textAfter, ...]
 * where:
 * - `pairedIndex` is index of paired element (undef for unpaired)
 * - `children` are children of paired element (undef for unpaired)
 * - `unpairedIndex` is index of unpaired element (undef for paired)
 * - `textAfter` is string after all elements (empty string, if there's nothing)
 *
 * `parts` length is always a multiple of 4
 *
 * Returns: Array<[elementIndex, children, after]>
 */
function getElements(
  parts: string[]
): Array<readonly [string | undefined, string, string | undefined]> {
  if (!parts.length) return []

  const [paired, children, unpaired, after] = parts.slice(0, 4)

  const triple = [paired || unpaired, children || "", after] as const
  return [triple].concat(getElements(parts.slice(4, parts.length)))
}

/**
 * `formatElements` - parse string and return tree of vue elements
 *
 * `value` is string to be formatted with <tag>Paired<tag/> or <tag/> (unpaired)
 * placeholders. `elements` is a array of vue slots which indexes
 * correspond to element indexes in formatted string
 */
export function formatElements(
  value: string,
  elements: { [key: string]: Slot | undefined } = {}
): Array<Node> {
  const parts = value.replace(nlRe, "").split(tagRe)

  // no inline elements, return
  if (parts.length === 1) return [value]

  const tree: Array<Node> = []

  const before = parts.shift()
  if (before) tree.push(before)

  for (const [index, children, after] of getElements(parts)) {
    const slot = typeof index !== "undefined" ? elements[index] : undefined
    let element: Element

    if (!slot) {
      console.error(
        `Can't use slot at index '${index}' as it is not declared in the original translation`
      )
      // ignore problematic element but push its children and elements after it
      element = h("span", children)
    } else {
      // slots display props with text interpolation
      // only way to do recursive thing is to give a component that will render
      // our subchildren / subslot
      const childrenInComponent = {
        setup: () => () => formatElements(children, elements),
      }
      element = slot({
        children: childrenInComponent,
      })
    }

    tree.push(element)

    if (after) tree.push(after)
  }

  return tree
}
