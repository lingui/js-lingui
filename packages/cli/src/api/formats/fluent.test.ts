import fs from "fs"
import path from "path"
import mockFs from "mock-fs"
import { mockConsole } from "@lingui/jest-mocks"

import format from "./fluent"

describe("fluent format", function() {
  afterEach(() => {
    mockFs.restore()
  })

  it("should write catalog in fluent format", function() {
    mockFs({
      locale: {
        en: mockFs.directory()
      }
    })

    const filename = path.join("locale", "en", "messages.flt")
    const catalog = {
      static: {
        translation: "Static message"
      },
      withOrigin: {
        translation: "Message with origin",
        origin: [["src/App.js", 4]]
      },
      withMultipleOrigins: {
        translation: "Message with multiple origin",
        origin: [["src/App.js", 4], ["src/Component.js", 2]]
      },
      withDescription: {
        translation: "Message with description",
        description: "Description is comment from developers to translators"
      },
      withComments: {
        comments: ["Translator comment", "This one might come from developer"],
        translation: "Support translator comments separately"
      },
      obsolete: {
        translation: "Obsolete message",
        comments: ["Old message"],
        obsolete: true
      },
      withFlags: {
        flags: ["fuzzy", "otherFlag"],
        translation: "Keeps any flags that are defined"
      },
      multilineString: {
        translation:
          "One morning, when Gregor Samsa woke from troubled dreams, he found himself" +
          "\ntransformed in his bed into a horrible vermin. He lay on his armour-like" +
          "\nback, and if he lifted his head a little he could see his brown belly," +
          "\nslightly domed and divided by arches into stiff sections. The bedding was" +
          "\nhardly able to cover it and seemed ready to slide off any moment. His many" +
          "\nlegs, pitifully thin compared with the size of the rest of him, waved about" +
          "\nhelplessly as he looked. \"What's happened to me?\" he thought. It wasn't" +
          "\na dream. His room, a proper human" +
          "\n  Two spaces before" +
          "\n     Four spaces before"
      }
    }

    format.write(filename, catalog, { language: "en" })
    const fltfile = fs.readFileSync(filename).toString()
    expect(fltfile).toMatchSnapshot()
  })

  it("should read catalog in fluent format", function() {
    const flt = fs
      .readFileSync(
        path.join(path.resolve(__dirname), "fixtures", "messages.flt")
      )
      .toString()

    mockFs({
      locale: {
        en: {
          "messages.flt": flt
        }
      }
    })

    const filename = path.join("locale", "en", "messages.flt")
    const actual = format.read(filename)
    expect(actual).toMatchSnapshot()
  })

  it("should write the same catalog as it was read", function() {
    const flt = fs
      .readFileSync(
        path.join(path.resolve(__dirname), "fixtures", "messages.flt")
      )
      .toString()

    mockFs({
      locale: {
        en: {
          "messages.flt": flt
        }
      }
    })

    const filename = path.join("locale", "en", "messages.flt")
    const catalog = format.read(filename)
    format.write(filename, catalog)
    const actual = fs.readFileSync(filename).toString()
    expect(actual).toEqual(flt)
  })
})
