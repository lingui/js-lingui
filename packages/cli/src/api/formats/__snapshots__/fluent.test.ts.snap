// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fluent format should read catalog in fluent format 1`] = `
Object {
  multilineString: Object {
    comments: undefined,
    defaults: undefined,
    description: undefined,
    flags: undefined,
    obsolete: false,
    origin: Array [],
    translation: One morning, when Gregor Samsa woke from troubled dreams, he found himself
transformed in his bed into a horrible vermin. He lay on his armour-like
back, and if he lifted his head a little he could see his brown belly,
slightly domed and divided by arches into stiff sections. The bedding was
hardly able to cover it and seemed ready to slide off any moment. His many
legs, pitifully thin compared with the size of the rest of him, waved about
helplessly as he looked. "What's happened to me?" he thought. It wasn't
a dream. His room, a proper human
  Two spaces before
     Four spaces before,
  },
  obsolete: Object {
    comments: Array [
      Old message,
    ],
    defaults: undefined,
    description: undefined,
    flags: undefined,
    obsolete: true,
    origin: Array [],
    translation: Obsolete message,
  },
  static: Object {
    comments: undefined,
    defaults: undefined,
    description: undefined,
    flags: undefined,
    obsolete: false,
    origin: Array [],
    translation: Static message,
  },
  withComments: Object {
    comments: Array [
      Translator comment,
      This one might come from developer,
    ],
    defaults: undefined,
    description: undefined,
    flags: undefined,
    obsolete: false,
    origin: Array [],
    translation: Support translator comments separately,
  },
  withDescription: Object {
    comments: undefined,
    defaults: undefined,
    description: Description is comment from developers to translators,
    flags: undefined,
    obsolete: false,
    origin: Array [],
    translation: Message with description,
  },
  withFlags: Object {
    comments: undefined,
    defaults: undefined,
    description: undefined,
    flags: Array [
      fuzzy,
      otherFlag,
    ],
    obsolete: false,
    origin: Array [],
    translation: Keeps any flags that are defined,
  },
  withMultipleOrigins: Object {
    comments: undefined,
    defaults: undefined,
    description: undefined,
    flags: undefined,
    obsolete: false,
    origin: Array [
      Array [
        src/App.js,
        4,
      ],
      Array [
        src/Component.js,
        2,
      ],
    ],
    translation: Message with multiple origin,
  },
  withOrigin: Object {
    comments: undefined,
    defaults: undefined,
    description: undefined,
    flags: undefined,
    obsolete: false,
    origin: Array [
      Array [
        src/App.js,
        4,
      ],
    ],
    translation: Message with origin,
  },
}
`;

exports[`fluent format should write catalog in fluent format 1`] = `
static = Static message

# #: src/App.js:4
withOrigin = Message with origin

# #: src/App.js:4
# #: src/Component.js:2
withMultipleOrigins = Message with multiple origin

# Description is comment from developers to translators
withDescription = Message with description

# #. Translator comment
# #. This one might come from developer
withComments = Support translator comments separately

# #~ # #. Old message
# #~ obsolete = Obsolete message

# #, fuzzy,otherFlag
withFlags = Keeps any flags that are defined

multilineString =
    One morning, when Gregor Samsa woke from troubled dreams, he found himself
    transformed in his bed into a horrible vermin. He lay on his armour-like
    back, and if he lifted his head a little he could see his brown belly,
    slightly domed and divided by arches into stiff sections. The bedding was
    hardly able to cover it and seemed ready to slide off any moment. His many
    legs, pitifully thin compared with the size of the rest of him, waved about
    helplessly as he looked. "What's happened to me?" he thought. It wasn't
    a dream. His room, a proper human
      Two spaces before
         Four spaces before

`;
